In [1]:
# Import libraries that we need
import numpy as np  # library to support large multi-dimensional arrays and matrices and functions     
import pandas as pd    # for data manipulation and analysis
import seaborn as sns    # plotting library
import matplotlib.pyplot as plt # plotting library

In [48]:
# Load dataset
df=pd.read_csv('./Blood biomarker baseline data_ADNI.csv')

C:\Users\nmupadhyay\AppData\Local\Temp\ipykernel_18216\386932887.py:2: DtypeWarning: Columns (0,1,3,7,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,25

In [49]:
df

Unnamed: 0   Unnamed: 1 Unnamed: 2  \
0    RID at enrollment (Participant roster ID)  Age\n(year)     Gender   
1                                            2         74.3       Male   
2                                            3         81.3       Male   
3                                            4         67.5       Male   
4                                            5         73.7       Male   
..                                         ...          ...        ...   
815                                       1425         75.6     Female   
816                                       1426         83.4       Male   
817                                       1427         69.6     Female   
818                                       1430         83.4     Female   
819                                       1435         82.4       Male   

    Unnamed: 3           Unnamed: 4        Unnamed: 5       Unnamed: 6  \
0    Education  Ethnicity\ncategory  Race subcategory  Married\nstatus   
1           16      Not Hisp/Latino             White          Married   
2           18      Not Hisp/Latino             White          Married   
3           10          Hisp/Latino             White          Married   
4           16      Not Hisp/Latino             White          Married   
..         ...                  ...               ...              ...   
815         16      Not Hisp/Latino             White         Divorced   
816         18      Not Hisp/Latino             White          Married   
817         12      Not Hisp/Latino             White          Widowed   
818         12      Not Hisp/Latino             White          Widowed   
819         12      Not Hisp/Latino             White          Married   

                Unnamed: 7  Unnamed: 8 APOGEE  ...  \
0    APOE4 genotype result  Diagnosis   L1CAM  ...   
1                        0          CN    NaN  ...   
2                        1          AD   0.73  ...   
3                        0        LMCI    NaN  ...   
4                        0          CN    NaN  ...   
..                     ...         ...    ...  ...   
815                      0        LMCI    NaN  ...   
816                      0        LMCI    NaN  ...   
817                      0        LMCI    NaN  ...   
818                      1          AD    NaN  ...   
819                      1          AD    NaN  ...   

                         Unnamed: 1286                Unnamed: 1287  \
0    TUDCA (Tauroursodeoxycholic acid)  UDCA (Ursodeoxycholic acid)   
1                                0.002                          NaN   
2                                0.005                        0.022   
3                                0.009                         0.03   
4                                0.024                         0.06   
..                                 ...                          ...   
815                              0.041                         0.11   
816                              0.002                          NaN   
817                              0.009                        0.079   
818                                NaN                          NaN   
819                              0.005                          NaN   

    Purines Data from file:  ADMCBarcelonapurine Unnamed: 1289 Unnamed: 1290  \
0                                    L.Glutamine       Taurine       Glycine   
1                                    530.6025076   35.49789484   712.2334028   
2                                    294.8877939   35.25381503   319.9505901   
3                                    355.4172375     68.622106   419.9529924   
4                                     435.975704   33.97121364   571.4948962   
..                                           ...           ...           ...   
815                                   391.240712     57.181786    219.929946   
816                                   390.011146     51.615926    421.982835   
817                                 

In [50]:
#Make second row as the header as 1st row has name of combined columns instead of individual column name. 

df.columns = df.iloc[0]

# Drop the second row (which is now the column headers)
df = df[1:]

# Reset the index if needed
df = df.reset_index().drop(columns=['index'])

df

0   RID at enrollment (Participant roster ID) Age\n(year)  Gender Education  \
0                                           2        74.3    Male        16   
1                                           3        81.3    Male        18   
2                                           4        67.5    Male        10   
3                                           5        73.7    Male        16   
4                                           6        80.4  Female        13   
..                                        ...         ...     ...       ...   
814                                      1425        75.6  Female        16   
815                                      1426        83.4    Male        18   
816                                      1427        69.6  Female        12   
817                                      1430        83.4  Female        12   
818                                      1435        82.4    Male        12   

0   Ethnicity\ncategory Race subcategory Married\nstatus  \
0       Not Hisp/Latino            White         Married   
1       Not Hisp/Latino            White         Married   
2           Hisp/Latino            White         Married   
3       Not Hisp/Latino            White         Married   
4       Not Hisp/Latino            White         Married   
..                  ...              ...             ...   
814     Not Hisp/Latino            White        Divorced   
815     Not Hisp/Latino            White         Married   
816     Not Hisp/Latino            White         Widowed   
817     Not Hisp/Latino            White         Widowed   
818     Not Hisp/Latino            White         Married   

0   APOE4 genotype result Diagnosis  L1CAM  ...  \
0                       0         CN   NaN  ...   
1                       1         AD  0.73  ...   
2                       0       LMCI   NaN  ...   
3                       0         CN   NaN  ...   
4                       0       LMCI   NaN  ...   
..                    ...        ...   ...  ...   
814                     0       LMCI   NaN  ...   
815                     0       LMCI   NaN  ...   
816                     0       LMCI   NaN  ...   
817                     1         AD   NaN  ...   
818                     1         AD   NaN  ...   

0   TUDCA (Tauroursodeoxycholic acid) UDCA (Ursodeoxycholic acid)  \
0                               0.002                         NaN   
1                               0.005                       0.022   
2                               0.009                        0.03   
3                               0.024                        0.06   
4                               0.003                         NaN   
..                                ...                         ...   
814                             0.041                        0.11   
815                             0.002                         NaN   
816                             0.009                       0.079   
817                               NaN                         NaN   
818                             0.005                         NaN   

0    L.Glutamine      Taurine      Glycine      Inosine Hypoxanthine  \
0    530.6025076  35.49789484  712.2334028  1.862880977  1.325800862   
1    294.8877939  35.25381503  319.9505901  1.454096432  0.319595089   
2    355.4172375    68.622106  419.9529924  1.703996497  0.697515115   
3     435.975704  33.97121364  571.4948962  2.248941525  0.472482963   
4    290.5517349  34.74466791  340.8412289  0.220181672  0.246990993   
..           ...          ...          ...          ...          ...   
814   391.240712    57.181786   219.929946     0.877569     0.588239   
815   390.011146    51.615926   421.982835     1.987726     0.712408   
816   344.785668    30.799028   338.636078     0.537732     0.432964   
817   347.490026    33.378196   890.085233     0.047978     0.370343   
818   409.153789    71.449738   494.883447     0.120437     0.564498   

0       Xanthine   Xant

In [51]:
# Inspect: number of samples and features
df.shape

# Inspect: number of samples and number of samples
print(str("The dataset has ") + str(df.shape[0])+str(' samples / instances & ')+str(df.shape[1])+ str(' features'))

The dataset has 819 samples / instances & 1296 features


In [52]:
# Inspect the dataset

df.dtypes

0
RID at enrollment (Participant roster ID)    object
Age\n(year)                                  object
Gender                                       object
Education                                    object
Ethnicity\ncategory                          object
                                              ...  
Inosine                                      object
Hypoxanthine                                 object
Xanthine                                     object
Xanthosine                                   object
Uric.acid                                    object
Length: 1296, dtype: object

In [53]:
# Calculate the total number of samples (rows)
total_num_samples = df.shape[0]

total_num_samples

# Calculate the count of missing values in each column
num_missing = df.isnull().sum()

# Calculate the percentage of missing values for each column
percent_missing = ((num_missing / total_num_samples)* 100)

percent_missing

# Set a threshold for missing values (35% in this case)
threshold = 35

# Identify columns with missing values greater than the threshold
columns_to_drop = percent_missing[percent_missing > threshold].index

# Drop the identified columns from the DataFrame
df.drop(columns=columns_to_drop, inplace=True)

df

0   RID at enrollment (Participant roster ID) Age\n(year)  Gender Education  \
0                                           2        74.3    Male        16   
1                                           3        81.3    Male        18   
2                                           4        67.5    Male        10   
3                                           5        73.7    Male        16   
4                                           6        80.4  Female        13   
..                                        ...         ...     ...       ...   
814                                      1425        75.6  Female        16   
815                                      1426        83.4    Male        18   
816                                      1427        69.6  Female        12   
817                                      1430        83.4  Female        12   
818                                      1435        82.4    Male        12   

0   Ethnicity\ncategory Race subcategory Married\nstatus  \
0       Not Hisp/Latino            White         Married   
1       Not Hisp/Latino            White         Married   
2           Hisp/Latino            White         Married   
3       Not Hisp/Latino            White         Married   
4       Not Hisp/Latino            White         Married   
..                  ...              ...             ...   
814     Not Hisp/Latino            White        Divorced   
815     Not Hisp/Latino            White         Married   
816     Not Hisp/Latino            White         Widowed   
817     Not Hisp/Latino            White         Widowed   
818     Not Hisp/Latino            White         Married   

0   APOE4 genotype result Diagnosis    AB40  ...  \
0                       0         CN  153.7  ...   
1                       1         AD  263.3  ...   
2                       0       LMCI    192  ...   
3                       0         CN  246.1  ...   
4                       0       LMCI    289  ...   
..                    ...        ...    ...  ...   
814                     0       LMCI   31.8  ...   
815                     0       LMCI  153.2  ...   
816                     0       LMCI  152.4  ...   
817                     1         AD    NaN  ...   
818                     1         AD    NaN  ...   

0   TMCA.a.b. (Tauromuricholic acid) TUDCA (Tauroursodeoxycholic acid)  \
0                              0.006                             0.002   
1                              0.011                             0.005   
2                                NaN                             0.009   
3                              0.013                             0.024   
4                                NaN                             0.003   
..                               ...                               ...   
814                            0.006                             0.041   
815                            0.006                             0.002   
816                              NaN                             0.009   
817                            0.022                               NaN   
818                            0.025                             0.005   

0    L.Glutamine      Taurine      Glycine      Inosine Hypoxanthine  \
0    530.6025076  35.49789484  712.2334028  1.862880977  1.325800862   
1    294.8877939  35.25381503  319.9505901  1.454096432  0.319595089   
2    355.4172375    68.622106  419.9529924  1.703996497  0.697515115   
3     435.975704  33.97121364  571.4948962  2.248941525  0.472482963   
4    290.5517349  34.74466791  340.8412289  0.220181672  0.246990993   
..           ...          ...          ...          ...          ...   
814   391.240712    57.181786   219.929946     0.877569     0.588239   
815   390.011146    51.615926   421.982835     1.987726     0.712408   
816   344.785668    30.799028   338.636078     0.537732     0.432964   
817   347.490026    33.378196   890.085233     0.047978     0.370343   
818   409.153789    71.4

In [54]:
print(df.dtypes)

0
RID at enrollment (Participant roster ID)    object
Age\n(year)                                  object
Gender                                       object
Education                                    object
Ethnicity\ncategory                          object
                                              ...  
Inosine                                      object
Hypoxanthine                                 object
Xanthine                                     object
Xanthosine                                   object
Uric.acid                                    object
Length: 1225, dtype: object


In [55]:
import pandas as pd
from sklearn.impute import SimpleImputer

# Create a function to try to convert values to numeric, handling errors gracefully
def try_convert_numeric(value):
    try:
        return pd.to_numeric(value)
    except (ValueError, TypeError):
        return value  # Return the original value if conversion fails

# Apply the function to each column in the DataFrame
df = df.applymap(try_convert_numeric)

# Separate numerical and categorical columns
numerical_columns = df.select_dtypes(include=['number'])
categorical_columns = df.select_dtypes(exclude=['number'])

# Now, 'df' should have appropriate data types for each column

# Impute missing numerical values with the median
numerical_imputer = SimpleImputer(strategy='median')
df[numerical_columns.columns] = numerical_imputer.fit_transform(df[numerical_columns.columns])

In [57]:
columns_to_keep = ["AB40", "AB42", "PLASMA_NFL", "PLASMATAU", "Diagnosis "]
df = df[columns_to_keep]

df

0     AB40  AB42  PLASMA_NFL  PLASMATAU Diagnosis 
0    153.7  50.9        28.2       2.41         CN
1    263.3  58.1       188.3       3.84         AD
2    192.0  59.8        31.2       3.76       LMCI
3    246.1  46.7        36.3       3.81         CN
4    289.0  44.3        52.5       2.86       LMCI
..     ...   ...         ...        ...        ...
814   31.8  26.7        36.9       2.65       LMCI
815  153.2  38.4        36.9       2.65       LMCI
816  152.4  48.6        36.9       2.65       LMCI
817  153.8  36.6        69.0       2.54         AD
818  153.8  36.6        35.3       3.34         AD

[819 rows x 5 columns]

In [58]:
#Converting categorical variable to numerical variables through label encoder:

from sklearn.preprocessing import LabelEncoder

# Create an instance of LabelEncoder
label_encoder = LabelEncoder()

# List of column names you want to encode
columns_to_encode = ["Diagnosis "]


# Iterate through the columns you want to encode
for column in columns_to_encode:
    # Convert all values in the column to strings
    df[column] = df[column].astype(str)
    
    # Apply LabelEncoder
    df[column] = label_encoder.fit_transform(df[column])

    
# Now df contains encoded values

df

0     AB40  AB42  PLASMA_NFL  PLASMATAU  Diagnosis 
0    153.7  50.9        28.2       2.41           1
1    263.3  58.1       188.3       3.84           0
2    192.0  59.8        31.2       3.76           2
3    246.1  46.7        36.3       3.81           1
4    289.0  44.3        52.5       2.86           2
..     ...   ...         ...        ...         ...
814   31.8  26.7        36.9       2.65           2
815  153.2  38.4        36.9       2.65           2
816  152.4  48.6        36.9       2.65           2
817  153.8  36.6        69.0       2.54           0
818  153.8  36.6        35.3       3.34           0

[819 rows x 5 columns]

In [70]:
df["AB42_DIV_AB40"] = df["AB42"] / df["AB40"]

In [71]:
df

0     AB40  AB42  PLASMA_NFL  PLASMATAU  Diagnosis   AB42_DIV_AB40
0    153.7  50.9        28.2       2.41           1       0.331165
1    263.3  58.1       188.3       3.84           0       0.220661
2    192.0  59.8        31.2       3.76           2       0.311458
3    246.1  46.7        36.3       3.81           1       0.189760
4    289.0  44.3        52.5       2.86           2       0.153287
..     ...   ...         ...        ...         ...            ...
814   31.8  26.7        36.9       2.65           2       0.839623
815  153.2  38.4        36.9       2.65           2       0.250653
816  152.4  48.6        36.9       2.65           2       0.318898
817  153.8  36.6        69.0       2.54           0       0.237971
818  153.8  36.6        35.3       3.34           0       0.237971

[819 rows x 6 columns]

In [72]:
# Check for missing values in the DataFrame
missing_values = df.isna().sum()

# Display the missing values count for each column
print(missing_values)

0
AB40             0
AB42             0
PLASMA_NFL       0
PLASMATAU        0
Diagnosis        0
AB42_DIV_AB40    6
dtype: int64


In [75]:
df = df.fillna(0)

df

0     AB40  AB42  PLASMA_NFL  PLASMATAU  Diagnosis   AB42_DIV_AB40
0    153.7  50.9        28.2       2.41           1       0.331165
1    263.3  58.1       188.3       3.84           0       0.220661
2    192.0  59.8        31.2       3.76           2       0.311458
3    246.1  46.7        36.3       3.81           1       0.189760
4    289.0  44.3        52.5       2.86           2       0.153287
..     ...   ...         ...        ...         ...            ...
814   31.8  26.7        36.9       2.65           2       0.839623
815  153.2  38.4        36.9       2.65           2       0.250653
816  152.4  48.6        36.9       2.65           2       0.318898
817  153.8  36.6        69.0       2.54           0       0.237971
818  153.8  36.6        35.3       3.34           0       0.237971

[819 rows x 6 columns]

In [79]:
# Check for missing values in the DataFrame
missing_values = df.isna().sum()

# Display the missing values count for each column
print(missing_values)

0
AB40             0
AB42             0
PLASMA_NFL       0
PLASMATAU        0
Diagnosis        0
AB42_DIV_AB40    0
dtype: int64


In [87]:
# Identify infinite values in the DataFrame
is_infinite = np.isinf(df)

# Replace infinite values with a large finite value (e.g., 1e9)
df[is_infinite] = 1e9

# Cleaned Dataset

In [88]:
df.to_csv('highly predictive_ cleaned_dataset_ADNI_06.09.2023.csv', index=False)

In [89]:
from sklearn.model_selection import train_test_split

# Define the features (X) and target variable (y)
X = df.drop('Diagnosis ', axis=1)  # Replace 'target_column' with your target variable name
y = df['Diagnosis ']

X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

In [90]:
# confirm splitting
print(X_train.shape, X_temp.shape, X_val.shape, X_test.shape, y_train.shape, y_temp.shape, y_val.shape, y_test.shape)

(491, 5) (328, 5) (164, 5) (164, 5) (491,) (328,) (164,) (164,)


In [91]:
from sklearn.preprocessing import StandardScaler  

# Perform feature scaling on the training set
scaler = StandardScaler()  # You can use other scalers as well
X_train_scaled = scaler.fit_transform(X_train)

In [92]:
# Confirm scaling of the train data

pd.DataFrame(X_train_scaled, columns=[X.columns])

0        AB40      AB42 PLASMA_NFL PLASMATAU AB42_DIV_AB40
0   -0.852786 -0.345972  -0.742423 -0.972981     -0.078406
1    0.709852 -0.102190  -0.934299  0.405772     -0.078406
2    0.367580  0.463731  -0.137276 -0.082358     -0.078406
3    0.726151 -0.067364  -1.126175 -0.844526     -0.078406
4   -2.005919 -1.895727   3.193493 -1.675204     -0.078406
..        ...       ...        ...       ...           ...
486 -1.413053 -1.382044  -0.137276 -0.082358     -0.078406
487 -0.871122 -0.537515  -0.048718 -0.853090     -0.078406
488  0.098651  0.045820  -0.137276 -0.082358     -0.078406
489 -0.086747 -0.084777   1.092698  3.719919     -0.078406
490  1.443294  0.899056  -0.747343  0.080352     -0.078406

[491 rows x 5 columns]

# Model Development and Evalaution

In [93]:
# Scale the validation set using the same scaler
X_val_scaled = scaler.transform(X_val)

# Scale the test set using the same scaler
X_test_scaled = scaler.transform(X_test)

In [94]:
#Model training and Hyperparametere tuning

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

# Define the Random Forest model
rf_model = RandomForestClassifier(random_state=42)

# Define hyperparameters and their possible values for tuning
rf_param_grid = {
    'n_estimators': [100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Perform hyperparameter tuning using GridSearchCV
rf_grid_search = GridSearchCV(estimator=rf_model, param_grid=rf_param_grid, cv=5)
rf_grid_search.fit(X_train_scaled, y_train)

# Get the best hyperparameters
best_rf_params = rf_grid_search.best_params_

# Train the Random Forest model with the best hyperparameters
rf_best_model = RandomForestClassifier(random_state=42, **best_rf_params)
rf_best_model.fit(X_train_scaled, y_train)

RandomForestClassifier(min_samples_leaf=2, min_samples_split=10,
                       random_state=42)

In [95]:
#Model Validation:

from sklearn.metrics import accuracy_score, classification_report

# Make predictions on the validation set
rf_val_predictions = rf_best_model.predict(X_val_scaled)

# Evaluate the models
rf_val_accuracy = accuracy_score(y_val, rf_val_predictions)

print("Validation Set Performance:")
print(f"Random Forest Accuracy: {rf_val_accuracy:.2f}")


# You can also print classification reports for more detailed evaluation
print("Random Forest Classification Report:")
print(classification_report(y_val, rf_val_predictions))

Validation Set Performance:
Random Forest Accuracy: 0.55
Random Forest Classification Report:
              precision    recall  f1-score   support

           0       0.57      0.43      0.49        37
           1       0.44      0.43      0.44        51
           2       0.60      0.68      0.64        76

    accuracy                           0.55       164
   macro avg       0.54      0.52      0.52       164
weighted avg       0.55      0.55      0.54       164



In [96]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Make predictions on the test set
rf_test_predictions = rf_best_model.predict(X_test_scaled)

# Evaluate the Random Forest model on the test set
rf_test_accuracy = accuracy_score(y_test, rf_test_predictions)
rf_test_classification_report = classification_report(y_test, rf_test_predictions)
rf_confusion_matrix = confusion_matrix(y_test, rf_test_predictions)

print("Test Set Performance (Random Forest Model):")
print(f"Accuracy: {rf_test_accuracy:.2f}")
print("Classification Report:")
print(rf_test_classification_report)
print("Confusion Matrix:")
print(rf_confusion_matrix)

Test Set Performance (Random Forest Model):
Accuracy: 0.54
Classification Report:
              precision    recall  f1-score   support

           0       0.56      0.51      0.53        47
           1       0.39      0.49      0.43        39
           2       0.62      0.58      0.60        78

    accuracy                           0.54       164
   macro avg       0.52      0.52      0.52       164
weighted avg       0.55      0.54      0.54       164

Confusion Matrix:
[[24 11 12]
 [ 5 19 15]
 [14 19 45]]


# Result: 
The result confirms the features do provide an accuracy of about 54% against model validation performance of 55% however additional features will be required to improve the accuracy of diagnosis.